src/modules/c.rs

```diff
@@ -0,0 +1,222 @@
+use super::{Context, Module, RootModuleConfig};
+
+use crate::configs::c::CConfig;
+use crate::formatter::StringFormatter;
+use crate::formatter::VersionFormatter;
+
+/// Creates a module with the current c version
+pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+    let mut module = context.new_module("c");
+    let config: CConfig = CConfig::try_load(module.config);
+    let is_c_project = context
+        .try_begin_scan()?
+        .set_extensions(&config.detect_extensions)
+        .set_files(&config.detect_files)
+        .set_folders(&config.detect_folders)
+        .is_match();
+
+    if !is_c_project {
+        return None;
+    }
+
+    let parsed = StringFormatter::new(config.format).and_then(|formatter| {
+        let c_compiler_info = if config.format.contains("$compiler_name")
+            || config.format.contains("$compiler_version")
+        {
+            context.exec_cmds_return_first(&[
+                // the compiler is usually cc, and --version works on gcc and clang
+                &["cc", "--version"],
+                // but on some platforms gcc is installed as *gcc*, not cc
+                &["gcc", "--version"],
+                // for completeness, although I've never seen a clang that wasn't cc
+                &["clang", "--version"],
+            ])
+        } else {
+            None
+        };
+
+        formatter
+            .map_meta(|var, _| match var {
+                "symbol" => Some(config.symbol),
+                _ => None,
+            })
+            .map_style(|variable| match variable {
+                "style" => Some(Ok(config.style)),
+                _ => None,
+            })
+            .map(|variable| match variable {
+                "compiler_name" => {
+                    let c_compiler_info = &c_compiler_info.as_ref()?.stdout;
+
+                    let c_compiler = if c_compiler_info.contains("clang") {
+                        "clang"
+                    } else if c_compiler_info.contains("Free Software Foundation") {
+                        "gcc"
+                    } else {
+                        "Unknown compiler"
+                    };
+                    Some(c_compiler).map(Ok)
+                }
+                _ => None,
+            })
+            .map(|variable| match variable {
```

Comment:
These should share `map`:
```suggestion
```

Reply:
I tried that, but ...

```
error[E0308]: mismatched types
  --> src/modules/c.rs:79:25
   |
79 | /                         VersionFormatter::format_module_version(
80 | |                             module.get_name(),
81 | |                             c_compiler_info.split_whitespace().nth(3)?,
82 | |                             config.version_format,
83 | |                         )
84 | |                         .map(Ok)
   | |________________________________^ expected `&str`, found struct `std::string::String`
   |
   = note: expected enum `std::option::Option<Result<&str, string_formatter::StringFormatterError>>`
              found enum `std::option::Option<Result<std::string::String, _>>`
```

and I couldn't figure out how to resolve it.

Reply:
They both need to have the same type (`as_str` or `as_deref` on the `String` might work). If that does not work, you could try using [`Cow`](https://doc.rust-lang.org/std/borrow/enum.Cow.html).

Reply:
With this code ...

```
+                        let formatted: Cow<String> = Cow::Owned(
+                            VersionFormatter::format_module_version(
+                                module.get_name(),
+                                c_compiler_info.split_whitespace().nth(3)?,
+                                config.version_format,
+                            )?
+                        );
+                        Some(formatted.into_owned().as_str()).map(Ok)
```

The type problem seems to be fixed, but now the compiler complains about ...

```
error[E0515]: cannot return value referencing temporary value
  --> src/modules/c.rs:84:25
   |
84 |                         Some(formatted.into_owned().as_str()).map(Ok)
   |                         ^^^^^----------------------^^^^^^^^^^^^^^^^^^
   |                         |    |
   |                         |    temporary value created here
   |                         returns a value referencing data owned by the current function
```

Unless I've misunderstood the doco for Cow, that's exactly what `into_owned` should be doing. What am I missing?

Reply:
Cow allows you to return both Borrowed and Owned Values from the same closure. The borrowed valued you create here don't have sufficient lifetimes. At the very least, I would expect `formatted.into_owned()` to require a variable binding.

But the fix here is much simpler. Just add `.map(Cow::Borrowed)` after `Some(compiler_name)`  and `.map(Cow::Owned)` after the `VersionFormatter` call.

Reply:
Thanks!